### PR TITLE
stdenv: provide a deterministically built gcc

### PR DESCRIPTION
**Pull request by baloo**
###### Motivation for this change

This is a proposal to fix one of the last issue on the road to reproducibility of nixos.
There is some background information here: #108475
There is also some discussion on: #445

`gcc`, when built, will run multiple stages. It will use performance data and profiling of one of those compilation stages to inject optimizations on a later stage. The purpose of this is to optimize performance.
This renders the build nondeterministic and impure since it inject local behavior of the builder. I believe this is contrary to the principles of nix, and @edolstra thesis.

Furthermore, I'm not sure how optimizations made for an hydra builder would affect performance on any other machine. To that end, I chose to make the `gcc` used by `stdenv` deterministic but keep an profiled built `gcc` in the default packages (build derivations deterministically albeit a bit slower, but you can run an optimized `gcc` in your `nix-shell` (if you're doing development)).
Note: This could probably use some benchmarks here.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
How do other distros handle this i.e. debian?

**Comment by baloo** [comment]
From what I can read, they run the profiledbootstrap, but the automated reproducibility tests times out.

**Comment by baloo** [comment]
I'm not too sure what I'm missing with the mesa build. It builds fine here.

**Comment by tomberek** [comment]
Looks like Arch doesn't do use profiledbootstrap it in favor of reproducibility: https://bugs.archlinux.org/task/56856

Nix provides an incredibly robust means to have the default be reproducible and allow opt-in for optimization. This means we have more options in this arena.

**Comment by baloo** [comment]
Running the following benchmark:
https://gist.github.com/baloo/45cebcaa2057d1cbb8a875338ce4ecf3

Trying to compare build performance before/after such a change, I get the following results:
```
==================
building: bc
==== baseline ====
0.01user 0.00system 0:11.51elapsed 0%CPU (0avgtext+0avgdata 21296maxresident)k
0inputs+0outputs (0major+1332minor)pagefaults 0swaps
==== target ====
0.01user 0.01system 0:12.70elapsed 0%CPU (0avgtext+0avgdata 21196maxresident)k
0inputs+0outputs (0major+1335minor)pagefaults 0swaps
==================
==================
building: zsh
==== baseline ====
0.02user 0.02system 1:07.12elapsed 0%CPU (0avgtext+0avgdata 21276maxresident)k
0inputs+0outputs (0major+1336minor)pagefaults 0swaps
==== target ====
0.03user 0.00system 1:11.78elapsed 0%CPU (0avgtext+0avgdata 21216maxresident)k
0inputs+0outputs (0major+1333minor)pagefaults 0swaps
==================
```

I'd need to run larger builds to compare (only ran zsh and bc for now)

**Comment by baloo** [comment]
Note: there is an issue with gfortran9 introduced here. Would only build on my second attempt. Not too sure what happened.

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 112928` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>26 packages built:</summary>
  <ul>
    <li>gcc</li>
    <li>gcc-arm-embedded (gcc-arm-embedded-10)</li>
    <li>gcc-arm-embedded-6</li>
    <li>gcc-arm-embedded-7</li>
    <li>gcc-arm-embedded-8</li>
    <li>gcc-arm-embedded-9</li>
    <li>gccForLibs (gcc-unwrapped)</li>
    <li>gcc10 (gcc_latest)</li>
    <li>gcc10Stdenv</li>
    <li>gcc48</li>
    <li>gcc49</li>
    <li>gcc49Stdenv</li>
    <li>gcc6</li>
    <li>gcc6Stdenv</li>
    <li>gcc7</li>
    <li>gcc7Stdenv</li>
    <li>gcc8</li>
    <li>gcc8Stdenv</li>
    <li>gcc9</li>
    <li>gcc9Stdenv</li>
    <li>gccMultiStdenv</li>
    <li>gccStdenv</li>
    <li>gccStdenvNoLibs</li>
    <li>gcc_debug</li>
    <li>gcc_multi</li>
    <li>gccgo (gccgo6)</li>
  </ul>
</details>

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 112928` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>gfortran (gfortran9)</li>
    <li>gfortran10</li>
    <li>gfortran7</li>
    <li>gfortran8</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>gfortran48</li>
    <li>gfortran49</li>
    <li>gfortran6</li>
  </ul>
</details>

**Comment by baloo** [comment]
@tomberek Yes I hit the same issues as well, I'm not too sure what went wrong with the gfortran evaluation.

**Comment by tomberek** [comment]
Running @baloo's benchmark I get around 2-5% slowdown.

An interesting question is how much slowdown is acceptable. If reproducible builds meant a 100% slowdown, we'd have a much harder time justifying it. Having robust reproducible builds allows a kind of parallelism in the package building ecosystem that would provide an order of magnitude speed up in how fast channels update, allow for greater coordination between mutually distrusted builders, easier regression testing, etc. So where would we draw the line? 10%? Has any of the reproducible-builds community set some rules of thumb for the effort?

**Comment by baloo** [comment]
(sorry to the last people I pulled in, in switching from staging to staging-next, I got confused with the naming of branches there :x)

**Comment by r-burns** [comment]
Just spitballing, would it be possible to:
* pregenerate profiling data, possibly nondeterministically
* upload it to tarballs.nixos and use it as a fixed-output input to the gcc build

?

Presumably this would make gcc deterministic with no loss of performance, at the cost of more complicated infrastructure.

**Comment by baloo** [comment]
A couple things here:
 - This would introduce a significant amount of work for the maintainer with each version of gcc, as the profile would need to be regenerated for each version of gcc (and arguably for each dependency tree change of the gcc build)
 - This inherently goes against some of the principles of reproducible-builds:
    The motivation behind the Reproducible Builds project is therefore to allow verification that no vulnerabilities or backdoors have been introduced during this compilation process. By promising identical results are always generated from a given source, this allows multiple third parties to come to a consensus on a “correct” result, highlighting any deviations as suspect and worthy of scrutiny.
    https://reproducible-builds.org/

That being said, if that's the road nixos want to take, I can have a look.

**Comment by r-burns** [comment]
Ah yeah interesting. I assumed the profile data would be human-readable so you don't have to trust it. If that's not the case and gcc doesn't parse/sanitize the profile data, then you have no protection from adversarial input. Either way, if it has to be regenerated for every version bump, then that's certainly more effort than anyone would be willing to put in.

Anyway I'd assume the 5% hit is well worth it for reproducibility. Just wanted to put the profile seed idea out there for consideration.

**Comment by tomberek** [comment]
Both approaches are worth considering. In the interest of incremental changes we can proceed with this change (take the perf hit on unstable, learn about any other perf consequences) and explore the “pre-seeded profiling” idea as a follow-on.

**Comment by baloo** [comment]
@r-burns as far as I can tell gcc stores those profile data in `.gcda` files

here is an example of such a file (hexdumped):
https://gist.github.com/baloo/8e9285bfdf06ea34687f3c0e59cf9130

I don't believe those are meant to be human readable.
A single build of gcc produces 41MB of such files.

**Comment by Gaelan** [comment]
The other important question in my mind is whether it's possible to force a mis-compilation with invalid profile data. If the worse-case scenario from malicious profile data is a slower compiler, that's not too bad.

**Comment by baloo** [comment]
If I read the code correctly, that should not be possible. The gcov parser just feeds a call count to each branches/edges.
This is fed in the `gcc/mcf.c` to adjust the weight of each edge of the call graph.

Beside throwing the call graph out of whack I don't believe this could create incorrect programs.

But I'm not a gcc developer, so take that with a grain of salt.

**Comment by FRidh** [comment]
I am of the opinion we should make all gcc provided reproducible, just like every other package in Nixpkgs. If users want more performance, they can recompile gcc with profiling enabled, with `march=native`, and whatever other optimizations they would like to enable.

**Comment by baloo** [comment]
@fridh so, just to be clear. You mean setting `profiledCompiler` to false by default, and leave it up to the end-user to override that in an overlay or something. Is that correct?

**Comment by baloo** [comment]
Also I was think about adding a `reproducibleBuild ? true` parameter to the gcc closure along with an assert like
```
# profiledCompiler builds inject non-determisn in the builds. If turned on, we can't provide reproducible builds anymore
assert reproducibleBuild -> profiledCompiler == false;
```
Just to ensure what the intention of the user of the closure is.

**Comment by FRidh** [comment]
@baloo right, I did the same in https://github.com/NixOS/nixpkgs/pull/107965/.

Ideally, we would not have a `reproducibleBuild` parameter, because it should always be, but indeed in order to be clear its needed for the assert.

**Comment by Mic92** [comment]
> I am of the opinion we should make all gcc provided reproducible, just like every other package in Nixpkgs. If users want more performance, they can recompile gcc with profiling enabled, with `march=native`, and whatever other optimizations they would like to enable.

I don't like the idea that I need to recompile my compiler.

**Comment by Mic92** [comment]
> @FRidh so, just to be clear. You mean setting `profiledCompiler` to false by default (and not just in stdenv), and leave it up to the end-user to override that in an overlay or something. Is that correct?
> 
> This current proposal, only affects the gcc used in the stdenv (so hopefully anything nix itself uses to build), and leaves the optimizations on for the rest (my idea was: if you're doing software development and you want local gcc to go as fast as possible and not affect your development cycle).

The state of the art is that most people use the gcc from stdenv as this is how shell.nix environments are built - many of them don't even know how to override the compiler from their. So you are making development cycles 10% slower for most people.

**Comment by Mic92** [comment]
To be clear I am not concerned about compilation speed in nixpkgs more about the interactive use-case.

**Comment by baloo** [comment]
the `gcc_latest` with optmizations wouldn't solve problems for people running `nix-shell` for development and using provided gcc.
Or am I missing something?

**Comment by Mic92** [comment]
> the `gcc_latest` with optmizations wouldn't solve problems for people running `nix-shell` for development and using provided gcc.
> Or am I missing something?

No we would also need something like `fastStdenv = overrideCC stdenv gcc_latest`. Or maybe move this over to https://github.com/numtide/devshell

**Comment by bmwiedemann** [comment]
If you look at the `gzip` example linked in https://github.com/bmwiedemann/theunreproduciblepackage/tree/master/pgo you find that there is a way to make Profile-Guided-Optimizations deterministic without shipping pre-recorded profiling data.
You 'only' need to ensure that everything that happens in the profiling run is 100% deterministic.
gcc's profiling run includes compiling of gcc, so it might be easier to replace that with a carefully crafted script, that covers most relevant code in a fully deterministic way.

**Comment by r-burns** [comment]
Tentatively milestoning this for 21.05 - I think all the concerns have been addressed here, and we are oh so close to a fully reproducible ISO :)

**Review by FRidh — APPROVED** [review]

**Comment by sternenseemann** [comment]
> If you look at the `gzip` example linked in https://github.com/bmwiedemann/theunreproduciblepackage/tree/master/pgo you find that there is a way to make Profile-Guided-Optimizations deterministic without shipping pre-recorded profiling data.
> You 'only' need to ensure that everything that happens in the profiling run is 100% deterministic.
> gcc's profiling run includes compiling of gcc, so it might be easier to replace that with a carefully crafted script, that covers most relevant code in a fully deterministic way.

We have something like that in nixpkgs actually: The [foot derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/terminal-emulators/foot/default.nix) supports PGO, but the build is made deterministic by generating the profiling data in the same way each time (by using a fixed seed for the random inputs generated for the profiling information collection). From my testing this makes the build deterministic as PGO doesn't seem to take timing information into account. I'm not sure however if the PGO build can also be reproducible or if the output would (deterministically) differ between different machines.

**Comment by baloo** [comment]
@sternenseemann just to see if I got things right.
foot is a terminal emulator and the stimulusGenerator is generating seeded random data as input to the emulator, and injects the profiling of the terminal back to the PGO build. Is that correct?

The way gcc does PGO is by compiling gcc itself and gathering profiling information from that build. I don't even know why there is entropy here.
Is that the non-deterministic build ordering (build is run in parallel, so different gcc calls are run out of order according to scheduling)? Is there any other factor?
Filesystem ordering? (gosh I fear the disorderfs now!)

**Comment by sternenseemann** [comment]
@baloo You got that right, yeah. `stimulusGenerator` generates a series of random inputs which are meant to be fed to the foot VTE parser for profiling purposes. Then an intermediate version of foot is compiled (with just the VTE parser so we don't need wayland) and that is fed with the inputs and generates profiling data.

That is then in turn used to compile the actual output binary.

I also need to correct myself, I checked again and the build is *actually reproducible* as well: I can reproduce the exact same foot as our hydra produced by using `nix-build --check -A foot` on two different x86_64-linux machines.

The upshot from this is: if we *could* track down the entropy in gcc's build we could have a reproducible gcc build which has PGO (unless I'm missing something). This is however probably quite the task, I wonder if upstream has some interest in this as well?

**Comment by baloo** [comment]
I'll a build with a patch like:
```
diff --git a/Makefile.in b/Makefile.in
index 36e369df6e7..060e97cd822 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -628,6 +628,11 @@ DEBUG_PREFIX_CFLAGS_FOR_TARGET = @DEBUG_PREFIX_CFLAGS_FOR_TARGET@

 XGCC_FLAGS_FOR_TARGET = $(FLAGS_FOR_TARGET) $(SYSROOT_CFLAGS_FOR_TARGET) $(DEBUG_PREFIX_CFLAGS_FOR_TARGET)

+ifeq ($(current_stage), STAGEtrain)
+# Ensure the PGO build is deterministic
+MAKE=$(MAKE) -j1
+endif
+
 # ------------------------------------
 # Miscellaneous targets and flag lists
 # ------------------------------------
```

**Comment by baloo** [comment]
Compiling twice on the same machine gets me the same result, but I still get variation between builds on two different machines. So it's not just the compilation ordering.

**Comment by raboof** [comment]
> Compiling twice on the same machine gets me the same result, but I still get variation between builds on two different machines. So it's not the compilation ordering

Making sure I'm understanding you correctly: you were testing with `-j1` and without the changes proposed in this PR, right?

So your experiment was intended to see if adding `-j1` would make gcc reproducible without the changes proposed in this PR, and it showed that `-j1` is not sufficient, and apparently there's still entropy coming from somewhere?

(I would like to see this PR merged, continuing work on adding PGO back in a deterministic way in parallel)

**Comment by tomberek** [comment]
```
[tom@tom:~/nixpkgs]$ nix store dump-path $(nix build .#legacyPackages.x86_64-linux.gcc10; nix path-info ./result) | md5sum
4f75cc04501df9828f30ac187341d966  -
```

**Comment by raboof** [comment]
> ```
> [tom@tom:~/nixpkgs]$ nix-store --dump $(nix-build -A gcc10.cc) | md5sum
> aa08b5e886bdd673feb9767e5f9bee99  -
> ```

I'm arriving at `aa08b5e886bdd673feb9767e5f9bee99` as well when building from this pr (specifically a961aeadae1abcce472014e3d0d7d34f83fe7685) :tada:

**Review by tomberek — APPROVED** [review]

**Comment by baloo** [comment]
> So your experiment was intended to see if adding `-j1` would make gcc reproducible without the changes proposed in this PR, and it showed that `-j1` is not sufficient, and apparently there's still entropy coming from somewhere?

Yes, exactly

**Comment by grahamc** [comment]
Does the profiling actually make a reasonable difference today? Aren't the benchmarks it is running actually running, for most users, on a very powerful datacenter grade machine which is also (generally) heavily loaded? Are these benchmarks useful for end users, who are most likely running GCC in a very different environment?

**Comment by sternenseemann** [comment]
> Does the profiling actually make a reasonable difference today? Aren't the benchmarks it is running actually running, for most users, on a very powerful datacenter grade machine which is also (generally) heavily loaded? Are these benchmarks useful for end users, who are most likely running GCC in a very different environment?

PGO — even though the profiling in principle has limited merit for the actual machines that we are building gcc for, impacts performance positively in a significant way. @baloo tested this further up in the thread <https://github.com/NixOS/nixpkgs/pull/112928#issuecomment-778508138>.

As far as I understand it, this is because profiling mostly reveals to the compiler, what codepaths are taken more often than others in *actual invocations* and thus allows further optimizatino.

**Review by raboof — APPROVED** [review]

**Review by tomberek — APPROVED** [review]

**Comment by raboof** [comment]
Looking through the last few months of discussion, it seems the rough consensus is that the performance hit looks acceptable (esp. with the availability of `fastStdenv`), so I think it's reasonable to merge this now.

Further research on how to "eat our cake and have it, too" will still be valuable, of course.

**Comment by vcunat** [comment]
I don't think `fastStdenv` really helps.  The majority of the extra cost will be paid by hydra.nixos.org, I believe.  I'm not sure if paying 7&ndash;12% extra is worth the benefit, but I personally provide for just an insignificant bit of the build farm...

**Comment by vcunat** [comment]
As for PGO generally, I believe the main benefit is in better estimates of which code-paths are hot and which are cold.  That should be mostly independent of the machine and more sensitive of the "training inputs".  The rule of thumb is to optimize hot paths for speed and cold paths for code size (which can also improve speed due to saving CPU instruction caches).

**Comment by raboof** [comment]
> That should be mostly independent of the machine and more sensitive of the "training inputs".

That's my expectation as well - if we can find a way to remove the current nondeterminism (adding `-j1` as tried in https://github.com/NixOS/nixpkgs/pull/112928#issuecomment-809714907 was not sufficient...) from the training we should definitely re-enable it.

**Comment by edolstra** [comment]
>  I'm not sure if paying 7–12% extra is worth the benefit, but I personally provide for just an insignificant bit of the build farm...

It's not just a cost for our build farm, it's a cost for every user of gcc. Slowing down every C/C++ build by 7-12% is a substantial price to pay...

**Comment by edolstra** [comment]
> and we are oh so close to a fully reproducible ISO :)

For a reproducible ISO, there is a simpler solution: don't include GCC in the ISO.

However from the perspective of content-addressable Nix, a deterministic GCC is certainly a lot better.

File: pkgs/top-level/all-packages.nix
Comment (edolstra) [inline pkgs/top-level/all-packages.nix:9824]: I don't think we should provide this attribute since 1) it's undiscoverable; 2) there is a risk that people start using it in nixpkgs (or elsewhere), and then we end up with closure bloat and libgcc/libstdc++ conflicts; 3) building gcc is expensive.

Comment (Mic92) [reply pkgs/top-level/all-packages.nix:9824]: I made it more discoverable and documented it [here](https://github.com/NixOS/nixpkgs/pull/120775) and [here](https://nixos.wiki/wiki/C#Faster_GCC_compiler)

**Pull request merged**
